### PR TITLE
[RFR] Inject store arguments in Behat hooks non step methods

### LINKED ISSUE #23: Sharing state fragments between non-step function
I need to share state fragments between two functions with annotation `@BeforeScenario`.
I can provide datas to store in the first function, but injection of datas in the second function doesn't seem to work.
```
  ┌─ @BeforeScenario  # V2\UserContext::buildUsers()
  │
  ╳  Warning: Missing argument 2 for V2\UserContext::buildUsers() in features/bootstrap/V2/UserContext.php line 120

```

### PR DESCRIPTION
**Pull request by vincentchalamon**
Fix #23

### CODE REVIEW

File: src/ScenarioStateCall.php [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+     * @param HookScope    $scope
+     * @param Hook         $hook
+     * @param null|integer $errorReportingLevel
+     */
+    public function __construct(HookScope $scope, Hook $hook, $arguments, $errorReportingLevel = null)
+    {
+        $environment;
+        $hook;
+        $arguments;
+        $errorReportingLevel;
```
Comment (gorghoa) [inline src/ScenarioStateCall.php]: I don’t quite see the point of those expressions 🤔

**Review by gorghoa — COMMENTED** [review]
thanks @vincentchalamon, this is very nice work 👍

File: src/ScenarioStateCall.php (hunk 1 above)
Comment (vincentchalamon) [reply src/ScenarioStateCall.php]: Hey, it's still WIP ;)

**Comment by vincentchalamon** [comment]
Unit tests in progress…

File: src/Hook/Dispatcher/ScenarioStateHookDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    /**
+     * Dispatches hooks for a specified event.
+     *
+     * @param HookScope $scope
+     *
+     * @return CallResults
+     */
+    public function dispatchScopeHooks(HookScope $scope)
+    {
+        $results = array();
```
Comment (meyerbaptiste) [inline src/Hook/Dispatcher/ScenarioStateHookDispatcher.php]: Use short syntax!

File: src/Hook/Dispatcher/ScenarioStateHookDispatcher.php [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            /** @var \ReflectionMethod $function */
+            $function = $hook->getReflection();
+
+            // No `@ScenarioStateArgument` annotation found
+            if (null === $this->reader->getMethodAnnotation($function, ScenarioStateArgument::class)) {
+                $results[] = $this->callCenter->makeCall(new HookCall($scope, $hook));
+                continue;
+            }
+
+            $paramsKeys = array_map(function($element) {
```
Comment (meyerbaptiste) [inline src/Hook/Dispatcher/ScenarioStateHookDispatcher.php]: What about `foreach` vs `array_map()`?

File: src/Hook/Tester/ScenarioStateHookableScenarioTester.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public function setUp(Environment $env, FeatureNode $feature, Scenario $scenario, $skip)
+    {
+        $setup = $this->baseTester->setUp($env, $feature, $scenario, true);
+
+        if ($skip) {
+            return $setup;
+        }
+
+        $hookCallResults = $this->dispatcher->dispatchScopeHooks(new BeforeScenarioScope($env, $feature, $scenario));
```
Comment (meyerbaptiste) [inline src/Hook/Tester/ScenarioStateHookableScenarioTester.php]: This var is useless.

File: src/Hook/Tester/ScenarioStateHookableScenarioTester.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public function tearDown(Environment $env, FeatureNode $feature, Scenario $scenario, $skip, TestResult $result)
+    {
+        $teardown = $this->baseTester->tearDown($env, $feature, $scenario, true, $result);
+
+        if ($skip) {
+            return $teardown;
+        }
+
+        $hookCallResults = $this->dispatcher->dispatchScopeHooks(new AfterScenarioScope($env, $feature, $scenario, $result));
```
Comment (meyerbaptiste) [inline src/Hook/Tester/ScenarioStateHookableScenarioTester.php]: This var is useless.

**Review by meyerbaptiste — CHANGES_REQUESTED** [review]

**Comment by vincentchalamon** [comment]
[Scrutinizer patches](https://goo.gl/MaBCXM) are wrong: http://www.php-fig.org/psr/psr-2/#closures
Some Scrutinizer issues are wrong (duplicate code, PHPUnit mocks), but the other issues should be fixed in another PR (https://goo.gl/cA4uEC)

**Review by gorghoa — APPROVED** [review]
Nice :) 👍

**Comment by gorghoa** [comment]
@meyerbaptiste, ok for you?

**Comment by gorghoa** [comment]
@gregcop1, If you have some time to spend, could you verify that the PR fixes your use case? (If not, just trust @vincentchalamon’s work and everything will be fine I’m sure)

**Comment by gorghoa** [comment]
Note. I’m okay with the small BC breaks due to namespace reorganization (the `Exception` chiefly).

I suggest to ship it to minor version anymay, even if it conflicts with semver. I expect no one will feel the pain since it’s deeply internal subject and the user base is still small ;)

**Review by meyerbaptiste — APPROVED** [review]

**Pull request merged**